### ISSUE TITLE
MGRS Performance Issue

### ISSUE THREAD

**Issue by empbarney** [missioncommand/emp3-android#113]
I have sent following to NASA:

What is the issue?

I have 38 Path objects each with two positions that are rendered on the map (full globe is visible). It takes the renderer approximately FOUR seconds to complete this task and application is non-responsive to user gestures in this interval (render is in on the UI thread). We want to know how this FOUR seconds interval can be reduced so that user doesn't see an unresponsive application. Depending on Camera position this interval can reach up to 7 to 8 seconds, I have not been able to gather exact data for that case.

Where Did I get the 38 paths from?

We have ported MGRS code to WW Android and that is where the Paths are generated from.

What am I sending you?

I have updated code from "PathFragment" class in the world-wind tutorials. I have added  code to generate the Path objects and add them to a layer. Code adds and removes the layer at an interval of 10 seconds. I am doing this repeatedly to see if performance changes on second, third iteration etc, it doesn't. In the Code you will see two private classes AddLayer and RemoveLayer that add and remove the layers. There is also an MGRSLayer class that has "doRender" overridden. It is in this overridden method where Path object are generated, added to the Layer and then "super.doRender" is invoked. It is the "super.doRender" method that takes 4 seconds to return.

Following is the Logcat for the application:

03-29 16:23:44.105 30242-30242/gov.nasa.worldwindx.tutorials D/PathsFragment: Add Layer
03-29 16:23:44.105 30242-30242/gov.nasa.worldwindx.tutorials D/PathsFragment: createMGRS
03-29 16:23:44.105 30242-30242/gov.nasa.worldwindx.tutorials D/PathsFragment: addRenderable
03-29 16:23:44.105 30242-30242/gov.nasa.worldwindx.tutorials D/PathsFragment: start super.doRender

HERE IS THE  GAP, it varies between 3.5 to 4.5 seconds.

03-29 16:23:47.735 30242-30242/gov.nasa.worldwindx.tutorials D/PathsFragment: done super.doRender
03-29 16:23:54.105 30242-30242/gov.nasa.worldwindx.tutorials D/PathsFragment: Remove Layer

Hardware:

Samsung, Galaxy Tab A, Model Number SM-T550
Android Version 6.0.1

**Comment by empbarney** [comment]
[PathsFragment (1).java.txt](https://github.com/missioncommand/emp3-android/files/880331/PathsFragment.1.java.txt)

**Comment by empbarney** [comment]
Here is the scenario.

  1. We are in steady state. MGRS grid is drawn.
  2. User gesture on the screen causes multiple NAVIGATOR_MOVED events followed by NAVIGATOR_STOPPED event.
  3. On receiving the NAVIGATOR_STOPPED event, MGRS grid lines are recalculated, that produces a set of Path objects (RHUMB style) and a set of Text objects.
  4. Rendering those Path objects takes considerable amount of time.

  Instead of asking you to look at all of this, I grabbed the Path objects that were created for a specific Camera location and incorporated them in the example I sent you. (I hope you have received that attachment)

In the example:

line 94: handler.postDelayed(new AddLayer(wwd), 5000); //STARTS THE LOOP
line 110: handler.postDelayed(new RemoveLayer(wwd), 10000); // CONTINUES THE LOOP
line 124: handler.postDelayed(new AddLayer(wwd), 10000); // CONTINUES THE LOOP

AddLayer adds the Layer and RemoveLayer removes the layer. Only reason I created the LOOP was to see if time taken to render is affected by doing it repeatedly. I could simply have executed AddLayer once and stopped.

you will see that
line 162: super.doRender(rc); // TAKES approximately FOUR seconds.

This is just one Camera Position and a set of Path objects that are created for that position. In other Camera positions set of Path objects and cardinality of the set changes and hence the time taken changes.
